libqtile/notify.py

```diff
@@ -96,9 +115,16 @@ def __init__(self, summary, body='', timeout=-1, hints=None, app_name='',
 
 class NotificationManager:
     def __init__(self):
-        self.notifications = []
-        self.callbacks = []
+        # Use OrderedDict to maintain order
+        self._notifications = OrderedDict()
+        # Use a dict for indexing for quick finding prev/next notification
+        self._notifications_idx = dict()
+        self._current_id = 0
+        self.callbacks_add = []
+        self.callbacks_delete = []
         self._service = None
+        # Use a lock to protect notification concurrency.
+        self._lock = RLock()
```

Comment:
Do we need all the `threading` stuff? Won't it all be fine without, running within qtile's async loop?

Reply:
Oh I knew that I was going to have this conversation :) . It is the only way I see that it will be concurrent. Let me elaborate a bit.
- add: Runs on a separate thread than `main_thread` by default. So you cannot avoid concurrency issues even with one subscriber.
- delete: Should run on a separate thread than `main_thread` for the same reasons as `add`.

Example:

Widget clicks delete and at the same time new notification arrives. Even if delete runs on main thread (which it should not), add runs on a separate thread so there should be concurrency so that variables/attributes do not get messed up.

And this a concurrency issue with just the widget as a subscriber. If you have more than one subscriber created by a user, that potentially itself runs on a separate thread so it does not block Qtile, things get more complicated without locking.

The implementation will not have a performance impact whatsoever with any Use Case, unless you benchmark it (which is not the point anyways) with 1000 notification access/add/delete per second with multiple subscribers.

Now one comment I have on this is that reads block when they don't have to be necessarily. We could 
- use a read write lock like [this](https://pypi.org/project/readerwriterlock/) to improve performance but it adds a  dependency on qtile.
- or I could even write it on my own (maybe not the best idea) using something like Semaphores. But in the end. this will provide slight to almost no performance boost in Use Cases where notifications are not bursting.

Let me know what you think.

Cheers